In [55]:
# Connecting to Google drive

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## **Data Understanding**

In [56]:
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

In [57]:
# Loading the dataset

file_path = "/content/drive/MyDrive/CRB_Data"

application_data= pd.read_csv(file_path + "/1. ApplicationData.csv")
#/content/drive/MyDrive/CRB_Data/1. ApplicationData.csv

application_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9492 entries, 0 to 9491
Data columns (total 16 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Application_ID                                 9492 non-null   object 
 1   Application_Date                               9492 non-null   object 
 2   Loan_Approval_Status                           9354 non-null   object 
 3   Age_at_Application                             9471 non-null   float64
 4   Marital_Status                                 7712 non-null   object 
 5   Gender                                         9488 non-null   object 
 6   Own_Rent_Home                                  8188 non-null   object 
 7   Occupation_Professional_Category               9361 non-null   object 
 8   Type_of_Business_Industry_of_Employment        7210 non-null   object 
 9   Employment_Status                              9357 

**Observations:**

1. There are 9492 records of the data.

2. There are 16 features: Application_ID, Application_Date, Loan_Approval_Status, Age_at_Application, Marital_Status, Gender, Own_Rent_Home,
Occupation_Professional_Category,Type_of_Business_Industry_of_Employment, Employment_Status, Time_at_Current_Employment_(Months),Time_Employed_with_Previous_Employer_(Months), Number_of_Dependents Loan_Purpose, Monthly_Income, Government_Employee

3. The data types are: float64(5), object(11)

In [58]:
contracts_data = pd.read_csv(file_path + "/2. ContractsData.csv")
contracts_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9374 entries, 0 to 9373
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Contract_ID                 9374 non-null   object 
 1   Application_ID              7598 non-null   object 
 2   Key                         9374 non-null   int64  
 3   Loan_Type                   9374 non-null   object 
 4   Contract_Start_Date         9374 non-null   object 
 5   Expected_Contract_End_Date  9374 non-null   object 
 6   Actual_Contract_End_Date    2593 non-null   object 
 7   Loan_Term                   9374 non-null   object 
 8   Total_Loan_Amount           9374 non-null   float64
dtypes: float64(1), int64(1), object(7)
memory usage: 659.2+ KB


**Observations:**

1. There are 9374 records of the data.

2. There are 9 features: Contract_ID, Application_ID, Key, Loan_Type, Contract_Start_Date, Expected_Contract_End_Date, Actual_Contract_End_Date Loan_Term, Total_Loan_Amount.

3. The data types are: float64(1), int64(1), object(7)

In [59]:
snapshot_data = pd.read_csv(file_path + "/3. ContractsSnapshotData.csv")
snapshot_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274371 entries, 0 to 274370
Data columns (total 6 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Contract_ID              274371 non-null  object 
 1   Snapshot Date            274371 non-null  object 
 2   Outstanding Amount       274371 non-null  float64
 3   Due Amount               274371 non-null  float64
 4   Number of Days Past Due  79158 non-null   float64
 5   Contract Status          274371 non-null  object 
dtypes: float64(3), object(3)
memory usage: 12.6+ MB


**Observations:**

1. There are 274371 records of the data.

2. There are 6 features: Contract_ID, Snapshot Date, Outstanding Amount, Due Amount, Number of Days Past Due, Contract Status.

3. The data types are: float64(3), object(3)

In [60]:
CRB_Data = pd.read_csv(file_path + "/4. CRB Data.csv")
CRB_Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6750 entries, 0 to 6749
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Application_ID    6750 non-null   object
 1   Application_Date  6750 non-null   object
 2   CRB Score         6750 non-null   int64 
 3   CRB Grade         6750 non-null   object
dtypes: int64(1), object(3)
memory usage: 211.1+ KB


**Observations:**

1. There are 6750 records of the data.

2. There are 4 features: Application_ID, Application_Date, CRB Score, CRB Grade

3. The data types are: int64(1), object(3)

In [61]:
account_data = pd.read_csv(file_path + "/5. Current and Savings Account Data.csv")
account_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857677 entries, 0 to 857676
Data columns (total 31 columns):
 #   Column                                  Non-Null Count   Dtype  
---  ------                                  --------------   -----  
 0   ID                                      857677 non-null  int64  
 1   Account_Number                          857677 non-null  object 
 2   Application ID                          857672 non-null  object 
 3   Month (YYYY-MM)                         857647 non-null  object 
 4   Open Date                               857647 non-null  object 
 5   Closed Date                             161748 non-null  object 
 6   Average Balance                         857647 non-null  float64
 7   Overdraft Limit                         753484 non-null  float64
 8   Number of Days in Debit during Month    857647 non-null  float64
 9   Number of Days in Credit during Month   857647 non-null  float64
 10  Maximum End of Day Balance              8576

**Observations:**

1. There are 857677 records of the data.

2. There are 31 features: 'ID', 'Account_Number', 'Application ID', 'Month (YYYY-MM)',
       'Open Date', 'Closed Date', 'Average Balance', 'Overdraft Limit',
       'Number of Days in Debit during Month',
       'Number of Days in Credit during Month', 'Maximum End of Day Balance',
       'Minimum End of Day Balance', 'Sum of Monthly Debit Transactions',
       'Sum of Monthly Credit Transactions', 'Last Transaction Date',
       'Sum of ATM Deposits', 'Number of ATM Deposits',
       'Sum of ATM Withdrawals', 'Number of ATM Withdrawals',
       'Sum of Cheque Debit Transactions',
       'Number of Cheque Debit Transactions',
       'Sum of Bounced Cheque transactions', 'Number of Bounced Cheques',
       'Sum of Debit Card POS Transactions',
       'Number of Debit Card POS Transactions',
       'Sum of Over the Counter Withdrawals',
       'Number of Over the Counter withdrawals', 'Account Type',
       'Account Status', 'Number of Joint Holders', 'SpshotMonth'

3. The data types are:  float64(21), int64(1), object(9)

## **Merge Datasets: Application_Data AND CRB_Data**

#### The first merge being the application_data and the CRB_Data, then based on the output given we pick the required columns resuling to merge1

In [62]:
import pandas as pd

# Merge application_data with CRB_Data
merged_df = pd.merge(application_data, CRB_Data, how='inner')

# Display the first few rows of the merged dataset
merged_df.head()

Application_ID Application_Date Loan_Approval_Status  Age_at_Application  \
0     C302100061        1/31/2018             Approved                32.0   
1     C302101329        7/31/2017             Approved                34.0   
2     C302101572        7/31/2017             Approved                32.0   
3     C302102730       12/31/2017             Approved                29.0   
4     C302102967        1/31/2018             Approved                34.0   

  Marital_Status  Gender Own_Rent_Home Occupation_Professional_Category  \
0      Unmarried  Female           Own                              NaN   
1      Unmarried    Male        Family                         Employee   
2      Unmarried    Male           Own                     Professional   
3      Unmarried  Female        Family                          Officer   
4        Married    Male           Own          Armed Forces Occupation   

  Type_of_Business_Industry_of_Employment Employment_Status  \
0                                     NaN         Full-Time   
1          Transportation & Communication         Full-Time   
2                           Civil Service         Full-Time   
3                                 Tourism         Full-Time   
4                                   Other         Full-Time   

   Time_at_Current_Employment_(Months)  \
0                                 49.0   
1                                135.0   
2                                145.0   
3                                 19.0   
4                                 60.0   

   Time_Employed_with_Previous_Employer_(Months)  Number_of_Dependents  \
0                                            NaN                   2.0   
1                                            NaN                   1.0   
2                                            NaN                   3.0   
3                                            NaN                   1.0   
4                                            NaN                   2.0   

         Loan_Purpose  Monthly_Income Government_Employee  CRB Score CRB Grade  
0               Other         2344.67               False        619         C  
1  Debt Consolidation         2195.41               False        627         C  
2  Debt Consolidation         4200.19                True        250         E  
3  Debt Consolidation         1252.05               False        622         C  
4               Other         3505.98                True        250         E

In [63]:
## The necessary columns from the merged_df that comprises of application_data and CRB_Data
merge1_df = merged_df.loc[:, ['Loan_Approval_Status', 'Employment_Status','Time_at_Current_Employment_(Months)', 'Number_of_Dependents','Monthly_Income','CRB Score', 'CRB Grade']]
merge1_df.head()

Loan_Approval_Status Employment_Status  Time_at_Current_Employment_(Months)  \
0             Approved         Full-Time                                 49.0   
1             Approved         Full-Time                                135.0   
2             Approved         Full-Time                                145.0   
3             Approved         Full-Time                                 19.0   
4             Approved         Full-Time                                 60.0   

   Number_of_Dependents  Monthly_Income  CRB Score CRB Grade  
0                   2.0         2344.67        619         C  
1                   1.0         2195.41        627         C  
2                   3.0         4200.19        250         E  
3                   1.0         1252.05        622         C  
4                   2.0         3505.98        250         E

## **Concationation: Current and Savings Account Data**

#### Appending Columns (Side-by-Side Merge), If you want to combine different columns from multiple DataFrames without a common key, use pd.concat() with axis=1:

In [64]:

# Select specific columns to concat
merged_df2 = pd.concat([account_data[['Average Balance', 'Overdraft Limit', 'Sum of Monthly Credit Transactions', 'Number of Bounced Cheques']],
                       snapshot_data[['Outstanding Amount', 'Number of Days Past Due', 'Contract Status']]], axis=1)


merged_df2.head()

Average Balance  Overdraft Limit  Sum of Monthly Credit Transactions  \
0          1986.09              0.0                             9685.72   
1            23.45              NaN                               23.45   
2          8120.71              0.0                                0.00   
3            25.26              NaN                               56.12   
4          6624.10              0.0                                0.00   

   Number of Bounced Cheques  Outstanding Amount  Number of Days Past Due  \
0                        0.0                0.00                      NaN   
1                        0.0              215.08                      NaN   
2                        0.0              215.08                      NaN   
3                        0.0              215.08                      NaN   
4                        0.0              215.08                      NaN   

  Contract Status  
0          Closed  
1          Active  
2          Active  
3          Active  
4          Active

## **ContractsData: Picked the necessary columns from the data**

In [65]:
# Selecting only the specified columns from the ContractsData since the application_data and Contracts id is different from the rest of the data
contracts_df = contracts_data[['Loan_Type', 'Loan_Term', 'Total_Loan_Amount']]

# Displaying the first few rows
contracts_df.head()


Loan_Type      Loan_Term  Total_Loan_Amount
0  Unsecured Personal Loan  60 - Month(s)           24529.89
1  Unsecured Personal Loan  60 - Month(s)           35085.79
2  Unsecured Personal Loan  60 - Month(s)           53564.88
3  Unsecured Personal Loan  48 - Month(s)           29073.86
4  Unsecured Personal Loan  48 - Month(s)           11723.33

In [66]:
## Combining all the three dataframes
combined_df = pd.concat([merge1_df,  merged_df2, contracts_df], axis=1)  # axis=1 for column-wise concatenation
combined_df.head()

Loan_Approval_Status Employment_Status  Time_at_Current_Employment_(Months)  \
0             Approved         Full-Time                                 49.0   
1             Approved         Full-Time                                135.0   
2             Approved         Full-Time                                145.0   
3             Approved         Full-Time                                 19.0   
4             Approved         Full-Time                                 60.0   

   Number_of_Dependents  Monthly_Income  CRB Score CRB Grade  Average Balance  \
0                   2.0         2344.67      619.0         C          1986.09   
1                   1.0         2195.41      627.0         C            23.45   
2                   3.0         4200.19      250.0         E          8120.71   
3                   1.0         1252.05      622.0         C            25.26   
4                   2.0         3505.98      250.0         E          6624.10   

   Overdraft Limit  Sum of Monthly Credit Transactions  \
0              0.0                             9685.72   
1              NaN                               23.45   
2              0.0                                0.00   
3              NaN                               56.12   
4              0.0                                0.00   

   Number of Bounced Cheques  Outstanding Amount  Number of Days Past Due  \
0                        0.0                0.00                      NaN   
1                        0.0              215.08                      NaN   
2                        0.0              215.08                      NaN   
3                        0.0              215.08                      NaN   
4                        0.0              215.08                      NaN   

  Contract Status                Loan_Type      Loan_Term  Total_Loan_Amount  
0          Closed  Unsecured Personal Loan  60 - Month(s)           24529.89  
1          Active  Unsecured Personal Loan  60 - Month(s)           35085.79  
2          Active  Unsecured Personal Loan  60 - Month(s)           53564.88  
3          Active  Unsecured Personal Loan  48 - Month(s)           29073.86  
4          Active  Unsecured Personal Loan  48 - Month(s)           11723.33

In [67]:
combined_df.describe()

Time_at_Current_Employment_(Months)  Number_of_Dependents  \
count                            13.000000             11.000000   
mean                             50.461538              1.181818   
std                              41.796362              0.873863   
min                              18.000000              0.000000   
25%                              26.000000              1.000000   
50%                              36.000000              1.000000   
75%                              49.000000              1.500000   
max                             145.000000              3.000000   

       Monthly_Income   CRB Score  Average Balance  Overdraft Limit  \
count       13.000000   13.000000    857647.000000         753484.0   
mean      2305.544615  554.692308       758.212527              0.0   
std        833.837225  139.593090      2829.181786              0.0   
min       1252.050000  250.000000    -30735.730000              0.0   
25%       1775.100000  572.000000        11.720000              0.0   
50%       2161.660000  619.000000        79.740000              0.0   
75%       2344.670000  627.000000       552.215000              0.0   
max       4200.190000  676.000000    281359.910000              0.0   

       Sum of Monthly Credit Transactions  Number of Bounced Cheques  \
count                       857529.000000              857647.000000   
mean                           858.311773                   0.000002   
std                           3532.513718                   0.001527   
min                              0.000000                   0.000000   
25%                              0.000000                   0.000000   
50%                              0.000000                   0.000000   
75%                            521.660000                   0.000000   
max                         336844.080000                   1.000000   

       Outstanding Amount  Number of Days Past Due  Total_Loan_Amount  
count        2.743710e+05             79158.000000        9374.000000  
mean         2.516123e+04                65.570606       24271.437439  
std          4.084894e+04               141.958344       23807.037921  
min         -7.087329e+04                 0.000000           0.000000  
25%          3.697200e+03                 0.000000        8620.267500  
50%          1.123487e+04                 5.000000       15115.240000  
75%          3.028738e+04                60.000000       30893.970000  
max          1.059941e+06              1765.000000      239137.160000

In [68]:
combined_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857677 entries, 0 to 857676
Data columns (total 17 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   Loan_Approval_Status                 13 non-null      object 
 1   Employment_Status                    13 non-null      object 
 2   Time_at_Current_Employment_(Months)  13 non-null      float64
 3   Number_of_Dependents                 11 non-null      float64
 4   Monthly_Income                       13 non-null      float64
 5   CRB Score                            13 non-null      float64
 6   CRB Grade                            13 non-null      object 
 7   Average Balance                      857647 non-null  float64
 8   Overdraft Limit                      753484 non-null  float64
 9   Sum of Monthly Credit Transactions   857529 non-null  float64
 10  Number of Bounced Cheques            857647 non-null  float64
 11  Outstanding A

In [69]:

# Checking the features and no. of records in the dataset

print("The number of records are : ", combined_df.shape[0])
print("The number of features are : ", combined_df.shape[1])
print("The list of features is : ", combined_df.columns)


The number of records are :  857677
The number of features are :  17
The list of features is :  Index(['Loan_Approval_Status', 'Employment_Status',
       'Time_at_Current_Employment_(Months)', 'Number_of_Dependents',
       'Monthly_Income', 'CRB Score', 'CRB Grade', 'Average Balance',
       'Overdraft Limit', 'Sum of Monthly Credit Transactions',
       'Number of Bounced Cheques', 'Outstanding Amount',
       'Number of Days Past Due', 'Contract Status', 'Loan_Type', 'Loan_Term',
       'Total_Loan_Amount'],
      dtype='object')


**Observations:**

1. There are 857677 records of the data.

2. There are 17 features:
       'Loan_Approval_Status', 'Employment_Status',
       'Time_at_Current_Employment_(Months)', 'Number_of_Dependents',
       'Monthly_Income', 'CRB Score', 'CRB Grade', 'Average Balance',
       'Overdraft Limit', 'Sum of Monthly Credit Transactions',
       'Number of Bounced Cheques', 'Outstanding Amount',
       'Number of Days Past Due', 'Contract Status', 'Loan_Type', 'Loan_Term',
       'Total_Loan_Amount'
       
3. The data types are:  float64(11), object(6)

## **Data Cleaning**

In [70]:
# Checking the columns' titles and datatypes
combined_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857677 entries, 0 to 857676
Data columns (total 17 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   Loan_Approval_Status                 13 non-null      object 
 1   Employment_Status                    13 non-null      object 
 2   Time_at_Current_Employment_(Months)  13 non-null      float64
 3   Number_of_Dependents                 11 non-null      float64
 4   Monthly_Income                       13 non-null      float64
 5   CRB Score                            13 non-null      float64
 6   CRB Grade                            13 non-null      object 
 7   Average Balance                      857647 non-null  float64
 8   Overdraft Limit                      753484 non-null  float64
 9   Sum of Monthly Credit Transactions   857529 non-null  float64
 10  Number of Bounced Cheques            857647 non-null  float64
 11  Outstanding A

In [71]:
# Get missing values count and percentage

missing_values = combined_df.isnull().sum()

missing_percentage = (combined_df.isnull().sum() / len(combined_df)) * 100

# Create a summary dataframe

missing_summary_df = pd.DataFrame({
    'Missing Values': missing_values,
    'Percentage Missing': missing_percentage})

# Sort by percentage missing
missing_summary = missing_summary_df.sort_values('Percentage Missing', ascending=False).round(0);

missing_summary

Missing Values  Percentage Missing
Number_of_Dependents                         857666               100.0
Loan_Approval_Status                         857664               100.0
Time_at_Current_Employment_(Months)          857664               100.0
Monthly_Income                               857664               100.0
CRB Score                                    857664               100.0
CRB Grade                                    857664               100.0
Employment_Status                            857664               100.0
Loan_Type                                    848303                99.0
Loan_Term                                    848303                99.0
Total_Loan_Amount                            848303                99.0
Number of Days Past Due                      778519                91.0
Outstanding Amount                           583306                68.0
Contract Status                              583306                68.0
Overdraft Limit                              104193                12.0
Sum of Monthly Credit Transactions              148                 0.0
Number of Bounced Cheques                        30                 0.0
Average Balance                                  30                 0.0

In [72]:


# Fill categorical columns with mode (most frequent value)
categorical_cols =combined_df.select_dtypes(include=['object']).columns
for col in categorical_cols:
    combined_df[col].fillna(combined_df[col].mode()[0], inplace=True)

# Fill numerical columns with median (better for skewed data)
numerical_cols = combined_df.select_dtypes(include=['number']).columns
for col in numerical_cols:
    combined_df[col].fillna(combined_df[col].median(), inplace=True)

# Check if there are any remaining missing values
remaining_missing = combined_df.isnull().sum().sum()

print("Remaining Missing Values:", remaining_missing)

Remaining Missing Values: 0


In [73]:
crb = combined_df.isnull().sum()
crb

Loan_Approval_Status                   0
Employment_Status                      0
Time_at_Current_Employment_(Months)    0
Number_of_Dependents                   0
Monthly_Income                         0
CRB Score                              0
CRB Grade                              0
Average Balance                        0
Overdraft Limit                        0
Sum of Monthly Credit Transactions     0
Number of Bounced Cheques              0
Outstanding Amount                     0
Number of Days Past Due                0
Contract Status                        0
Loan_Type                              0
Loan_Term                              0
Total_Loan_Amount                      0
dtype: int64

In [74]:
# Checking for duplicates
print("No. of duplicates records in the dataset : ",crb.duplicated().sum())


No. of duplicates records in the dataset :  16


In [75]:

data = crb.drop_duplicates().sum()
print("No. of duplicates records in the dataset : ",data)

No. of duplicates records in the dataset :  0


In [76]:
# Save DataFrame to a CSV file
final_data = combined_df.to_csv("CRB_DATA.csv", index=False)
print("DataFrame saved as 'CRB_DATA.csv'")

DataFrame saved as 'CRB_DATA.csv'


In [77]:
file_path = "/content/drive/MyDrive/CRB_Data"

CRB = pd.read_csv(file_path + "/CRB_DATA.csv")
# Checking the features and no. of records in the dataset

print("The number of records are : ", CRB.shape[0])
print("The number of features are : ", CRB.shape[1])
print("The list of features is : ", CRB.columns)

The number of records are :  857677
The number of features are :  17
The list of features is :  Index(['Loan_Approval_Status', 'Employment_Status',
       'Time_at_Current_Employment_(Months)', 'Number_of_Dependents',
       'Monthly_Income', 'CRB Score', 'CRB Grade', 'Average Balance',
       'Overdraft Limit', 'Sum of Monthly Credit Transactions',
       'Number of Bounced Cheques', 'Outstanding Amount',
       'Number of Days Past Due', 'Contract Status', 'Loan_Type', 'Loan_Term',
       'Total_Loan_Amount'],
      dtype='object')


In [78]:
CRB.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857677 entries, 0 to 857676
Data columns (total 17 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   Loan_Approval_Status                 857677 non-null  object 
 1   Employment_Status                    857677 non-null  object 
 2   Time_at_Current_Employment_(Months)  857677 non-null  float64
 3   Number_of_Dependents                 857677 non-null  float64
 4   Monthly_Income                       857677 non-null  float64
 5   CRB Score                            857677 non-null  float64
 6   CRB Grade                            857677 non-null  object 
 7   Average Balance                      857677 non-null  float64
 8   Overdraft Limit                      857677 non-null  float64
 9   Sum of Monthly Credit Transactions   857677 non-null  float64
 10  Number of Bounced Cheques            857677 non-null  float64
 11  Outstanding A

**Observations:**

1. There are 56408 records of the data.

2. There are 17 features:
       'Loan_Approval_Status', 'Employment_Status',
       'Time_at_Current_Employment_(Months)', 'Number_of_Dependents',
       'Monthly_Income', 'CRB Score', 'CRB Grade', 'Average Balance',
       'Overdraft Limit', 'Sum of Monthly Credit Transactions',
       'Number of Bounced Cheques', 'Outstanding Amount',
       'Number of Days Past Due', 'Contract Status', 'Loan_Type', 'Loan_Term',
       'Total_Loan_Amount'.

3. The Data types are:  float64(11), object(6)